src/device/usbd.c

```diff
@@ -316,66 +318,142 @@ void usbd_task( void* param)
 #endif
 }
 
-void tud_control_interface_control_complete_cb(uint8_t rhport, uint8_t interface, tusb_control_request_t const * const p_request) {
-    if (_usbd_dev.itf2drv[ interface ] < USBD_CLASS_DRIVER_COUNT)
+//--------------------------------------------------------------------+
+// Control Request Parser & Handling
+//--------------------------------------------------------------------+
+
+// This handles the actual request and its response.
+// return false will cause its caller to stall control endpoint
+static bool process_control_request(uint8_t rhport, tusb_control_request_t const * p_request)

```

Comment:
Why not have this in usbd_control.c?

Reply:
usbd_control.c is only used for carrying out control transfer: data, status and invoke complete callback. Parsing and making sense of request is part of usbd.